src/util/spscfifo.h

```diff
@@ -0,0 +1,234 @@
+class SpscFifo final {
+  public:
+    typedef T value_type;
+    typedef S size_type;
+
+    explicit SpscFifo(size_type capacity)
+            : m_size(capacity + 1),
+              m_data(new T[m_size]),
+              m_front(0),
+              m_back(0),
+              m_pfront(0),
+              m_cback(0) {
+        static_assert(std::is_unsigned<size_type>::value, "size_type must be unsigned");
+        DEBUG_ASSERT(capacity > 0);
+        DEBUG_ASSERT(m_size > 0); // no overflow
+        DEBUG_ASSERT(std::atomic_is_lock_free(&m_front));
+        DEBUG_ASSERT(std::atomic_is_lock_free(&m_back));
+    }
+    SpscFifo(const SpscFifo&) = delete;
+    SpscFifo(SpscFifo&&) = delete;
+    ~SpscFifo() {
+        delete[] m_data;
+    }
+
+    size_type capacity() const {
+        return m_size - 1;
+    }
+
+    ///////////////////////////////////////////////////////////////////
+    // Producer operations
+    ///////////////////////////////////////////////////////////////////
+
+    // Check how many elements are immediately available for reading.
+    // With min_count = 0 only relaxed memory ordering is applied
+    // when accessing atomic values.
+    size_type push_peek(size_type min_count = 0) const {
+        return acquire_writable(min_count);
+    }
+
+    // Non-blocking skip write operation, i.e. only advance the
+    // write position
+    size_type push_skip(size_type count = 1) {
+        auto writable = math_min(acquire_writable(count), count);
+        if (writable > 0) {
+            DEBUG_ASSERT(writable < m_size);
+            m_pfront = (m_pfront + writable) % m_size;
+            DEBUG_ASSERT(m_pfront < m_size);
+            m_front.store(m_pfront, std::memory_order_release);
+        }
+        DEBUG_ASSERT(writable <= count);
+        return writable;
+    }
+
+    // Non-blocking write operation (move)
+    size_type push(value_type* data, size_type count = 1) {
+        DEBUG_ASSERT(data || !count);
+        auto writable = math_min(acquire_writable(count), count);
+        if (writable > 0) {
+            DEBUG_ASSERT(writable < m_size);
+            for (size_type i = 0; i < writable; ++i) {
+                m_data[(m_pfront + i) % m_size] = std::move(data[i]);
+            }
+            m_pfront = (m_pfront + writable) % m_size;
+            DEBUG_ASSERT(m_pfront < m_size);
+            m_front.store(m_pfront, std::memory_order_release);
+        }
+        DEBUG_ASSERT(writable <= count);
+        return writable;
+    }
+
+    // Non-blocking write operation (copy)
+    size_type push(const value_type* data, size_type count = 1) {
+        DEBUG_ASSERT(data || !count);
+        auto writable = math_min(acquire_writable(count), count);
+        if (writable > 0) {
+            DEBUG_ASSERT(writable < m_size);
+            for (size_type i = 0; i < writable; ++i) {
+                m_data[(m_pfront + i) % m_size] = data[i];
+            }
+            m_pfront = (m_pfront + writable) % m_size;
```

Comment:
why we store here "relaxed" and "release" short after.

Reply:
I think we should make sure to read and write the atomics only once par fifo access and cache their values in a local variable during usage.

Reply:
`m_cback` and `m_pfront` are already non-shared, cached values. Both members are documented accordingly, that's the basic idea of the algorithm.

Reply:
It is IMHO necessary, to better understand the code. And to have a performance gain independent from the used comilerer. 

The atomic type gurantees that the value is actually written and not optimized away. So we must not use it if optimization is desired.

In this case I thing = is overloaded with load which defaults to a sequential store. 

It could be also interesting, if the new code actually is faster. Sometimes trivial quirks destroys a theoretical performance gain making this PR pointless. 

In this case a memory fence is IMHO negligible compared to calling a sequential constructor vs. to a vectorized memcpy.

Reply:
Again: `m_pfront` and `m_cback` are simple integers, not atomics. The atomics `m_front` and `m_back` are accessed through explicit load and store operations.

The use cases were this new FIFO is used only push or pop single messages into/from the queue, not for plain audio data. The memcpy is dangerous if those types are not PODs. Safety first. For PODs the compiler might be able to optimize the loops away. If this is prevented by the modulo operation then two separate loops could be used.

The existing code already contains a memory leak, because the detached alloc/free operations of the POD message type are used incorrectly. And this is only a minor bug.

Reply:
Ok, now I see the issue.
So we store the same value redundant in two variables?
That should be fixed.

Reply:
That's the point of the algorithm!!!